scrapy_poet/injection.py

```diff
@@ -197,6 +216,37 @@ async def item_factory(page: page_object_cls) -> item_cls:  # type: ignore[valid
 
         return mapping_fn
 
+    @staticmethod
+    def _get_dynamic_deps_factory(
+        dynamic_types: List[type],
+    ) -> Callable[..., DynamicDeps]:
+        """Return a function that creates a :class:`.DynamicDeps` instance from its args.
+
+        It takes instances of types from ``dynamic_types`` as args and returns
+        a :class:`.DynamicDeps` instance where keys are types and values are
+        corresponding args. It has correct type hints so that it can be used as
+        an ``andi`` custom builder.
+        """
+
+        # inspired by dataclasses._create_fn()
```

Comment:
1. It'd be good to add a link to the source code (with a commit hash?), because in latest Python there is no more dataclasses._create_fn. 
2. What do you think about testing the string generation part separately? To make it clear what kind of code we're generating for the execution.

Sorry, I just find this code tricky, and it'd be good to try to make it a bit more approachable, for the future :)

Reply:
It's really good then that I've read the 3.11 implementation, the 3.12 one is more complicated :)